### ISSUE TITLE
DB file size is too big

### ISSUE THREAD

**Issue by morder** [realm/realm-java#3822]
Hello. I have a problem with DB file size. Some of our users reports about DB file size > 300Mb
So, I have some tests with realm database.

Ok, let's have a simple test.
It's a main thread, no more writes in any other threads!

        Realm realm = RealmCreator.getRealmInstance(RealmCreator.Db.Test);
        for (int i = 0; i < 1000; i++) {           
            realm.beginTransaction();
            TestEvent event = createTestEvent();
            realm.copyToRealmOrUpdate(event);
            realm.commitTransaction();                       
            LOGGER.showDbStats("iteration " + i);
        }
        realm.close();

     private TestEvent createTestEvent() {
             Random random = new Random();
             final TestEvent row = new TestEvent();
             row.setObjId(UUID.randomUUID().toString());
             row.setField1(random.nextBoolean());
             row.setField2(new BigInteger(130, random).toString(32));
             row.setField3(random.nextInt());
             row.setField4(new BigInteger(130, random).toString(32));
             row.setField5(random.nextLong());
             return row;
         }
    public class TestEvent extends RealmObject {
        @Index
        @PrimaryKey
        @SerializedName("id")
        private String objId;
    
        private boolean field1;
        private String field2;
        private int field3;
        private String field4;
        private long field5;
    }

This is my logs:
_iteration 0: DB files size: Test: 4 096
...
iteration 211: DB files size: Test: 163 840
iteration 212: DB files size: Test: 180 224
...
iteration 313: DB files size: Test: 294 912
iteration 314: DB files size: Test: 327 680
...
iteration 606: DB files size: Test: 524 288
iteration 607: DB files size: Test: 524 288
iteration 608: DB files size: Test: 589 824
iteration 609: DB files size: Test: 589 824
iteration 610: DB files size: Test: 589 824
iteration 611: DB files size: Test: 589 824
iteration 612: DB files size: Test: 589 824
iteration 613: DB files size: Test: 589 824
iteration 614: DB files size: Test: 655 360
iteration 615: DB files size: Test: 655 360
...
iteration 999: DB files size: Test: 851 968_

After this operation my DB file size every time is different, for example 917 504 bytes or 786 432 or 983 040 ...

If I run
`
Realm.compactRealm(RealmCreator.getRealmConfig(RealmCreator.Db.Test));
`

my DB compacted size now is 122 880 bytes every time

**So, I have a question - whats happend with DB and how can I prevent DB growth in single thread application and in multithread? (without use compact or use it rarely)**

realm-gradle-plugin:2.2.0
android 5.1.1 Nexus 4
Realm sync feature enabled: no
Android Studio 2.2.2

**Comment by Zhuinden** [comment]
Don't create every single object in a new transaction

**Comment by morder** [comment]
In my app I can't collect all events to one big transaction because they appears in different periods of time.

Well, I have a test 

        Realm realm = RealmCreator.getRealmInstance(RealmCreator.Db.Test);
        realm.beginTransaction();
        for (int i = 0; i < 1000; i++) {
            synchronized (sync) {
                TestEvent event = createTestEvent();
                realm.copyToRealmOrUpdate(event);
            }
        }
        realm.commitTransaction();
        realm.close();
        LOGGER.showDbStats();
`

Logs:
_DB files size: Test: 2 359 296_

Wow, 2+Mb really? Why? I don't understand what happend with db

compacted to 122 880 bytes

**Comment by Zhuinden** [comment]
Well you created 1000 versions of the Realm, it's bound to get big o-o

Anyways, you could try opening and closing the Realm inside the for loop. I'm kinda curious what happens then. As long as you don't have any other Realm instances lingering around, anyways.

**Comment by morder** [comment]
Trying this

        for (int i = 0; i < 1000; i++) {
            Realm realm = RealmCreator.getRealmInstance(RealmCreator.Db.Test);
            realm.beginTransaction();

            TestEvent event = createTestEvent();
            realm.copyToRealmOrUpdate(event);

            realm.commitTransaction();
            realm.close();

            LOGGER.showDbStats("iteration " + i);
        }

Logs:
_...
E/REALM: could not create WeakRealmNotifier ALooper message pipe: Too many open files
iteration 902: DB files size: Test: 786 432
E/REALM: could not create WeakRealmNotifier ALooper message pipe: Too many open files
iteration 903: DB files size: Test: 851 968
...
E/REALM: could not create WeakRealmNotifier ALooper message pipe: Too many open files
iteration 999: DB files size: Test: 851 968_

Compact: 122 880 bytes

**Comment by Zhuinden** [comment]
Well, if you can't reduce the number of transactions, then you'll need to compact from time to time.

**Comment by kneth** [comment]
As mentioned earlier, it looks like you have a Realm instance on another thread. See https://realm.io/docs/java/latest/#large-realm-file-size and https://realm.io/docs/java/latest/#controlling-the-lifecycle-of-realm-instances

**Comment by morder** [comment]
kneth, I have tested on new empty project.

**Comment by kneth** [comment]
@morder Did you draw any conclusions from your test? Can you share your test project?

**Comment by Zhuinden** [comment]
@kneth my conclusion would be that if you execute many transactions in a for loop, then the previous versions seem to be retained, and not immediately released. Which is odd, because `syncIfNeeded()` ought to clear these versions and reclaim the space if no other Realm locks the versions.

I wonder if this is a change compared to previous behavior? I haven't ever tried saving 10000 objects in 10000 transactions, personally.

**Comment by morder** [comment]
@kneth I made a project 
https://github.com/morder/RealmDbGrowthTest
just look at logcat

main thread, 1000 iterations of one simple transaction

first start:
_...
iteration 998: DB files size: Test: 786 432
iteration 999: DB files size: Test: 786 432
DB files size: Test: 786 432 <---- different and big size
DB files size: Test: 122 880 (after compact)_

second start:
_iteration 998: DB files size: Test: 917 504
iteration 999: DB files size: Test: 917 504
DB files size: Test: 917 504 <---- different and big size
DB files size: Test: 122 880 (after compact)_

**Comment by kneth** [comment]
@morder Thanks for the sample project. 

Running with only one thread, I observe that the Realm file increases in size. Which is fine, as the app is adding objects. But I also observe that the size stay constant over a number of iterations, and the size after 1000 iterations are much larger than after compaction. 

Internally, Realm will allocate memory when you creates a new object. But instead of allocated only what is needed, Realm allocate a larger chunk. The reason is that the operating system (Linux in this case), performs much better if you allocates in chunks that are aligned with page size. Realm does not free memory since it might lead to memory fragmentation (and degrading the performance).

Most databases have to do vacuum periodically (see for example http://stackoverflow.com/questions/2128336/what-does-it-mean-to-vacuum-a-database). Realm is not different - we just call it compacting instead. Since compacting is a blocking operation, we don't want to do it behind the back of the app developer - and potentially block the UI thread.

**Comment by diegomontoya** [comment]
@kneth I would side with @morder that the size increase in this case is "unnatural" to the end-user. There are no delete, modify combos, or other readers holding on to different versions of the db to explain multi-versions and data gaps that blows up the db. 

There is what I and probably more users would expect based on the @morder's code.

1) Loop add lots of data. Assume total data plus index is 120KB. 
2) During the loop add, realm ran out space and need to extend db size and size realm core allocates storage in chunks so let's assume it grows in 256KB chunks every time it needs to add space. 

With above assumptions, for the first loop, the size before compaction should be at most 256KB before compaction, and not 750KB in this case.

For the second loop, the allocation is chunk of 256KB is added to the compacted 120KB in size so the final realm size should be around 378KB before compaction, not 900KB. 

The question is if there are usable free space within realm during transaction add, why is realm not re-using them and keep allocating space? There are no modify/deletes happening here that may cause un-recyclable data-gaps. It is a fairly simple sequential data append, as far as the end-user is concerned.

**Comment by Zhuinden** [comment]
@diegomontoya well the end user shouldn't be using 1000 transactions for 1000 items, but I do wonder why the versions created by the transactions essentially get retained while looping.

**Comment by diegomontoya** [comment]
@Zhuinden It is not not optimized to do 1000 transactions in a loop but this loop is also valid as it simulates a stream of small unbuffered/frequent log events. End-users can buffer to optimize but that reduces durability and add another layer of bugs.

**Comment by kneth** [comment]
I have created an issue at Realm Core as I believe it is where the problem can be solved.

**Comment by finnschiermer** [comment]
@morder: I have investigated the project you provide. Thanks a lot.

The file grows due to fragmentation. This is an effect which is pronounced when the file is small, but as you add more objects more of the free space becomes reusable. The following numbers are obtained without use of primary keys/index, though:
~~~
elements: cost/element: free space overhead:
1000      115 byte      737 byte     640%
2000      115 byte      197 byte     171%
5000      118 byte      118 byte     100%
10000    118 byte      52 byte       44%
20000    118 byte      39 byte       33%
~~~
In summary: the project you provided behaves as intended. My guess is that it doesn't really explain why some users report very large databases.

**Comment by diegomontoya** [comment]
@finnschiermer Is there a way for users to tune the realm-core to reduce this fragmentation? And can you explain a little bit about how this fragmentation is created by core in @morder's serial add only transaction code? Thanks.

**Comment by Zhuinden** [comment]
@diegomontoya (...I'm still not surprised he gets additional size increase if he creates a new version for each object that is written to the database)

**Comment by finnschiermer** [comment]
@diegomontoya: One very high level way of explaining it, is that the objects are not just merely added though that is how it looks. Internally they are "repackaged" into larger chunks which makes later access faster and over time also decreases memory usage. To ensure crash safety we cannot just recycle memory immediately, so repackaging involves both allocation and freeing but always such that we allocate first, then commit, then recycle memory in the following transaction. In this particular case, the pattern of allocation is worse than what we typically see.

At the moment we don't provide any way to tune it.

Just to avoid misunderstandings: the overhead I reported earlier depends not only on the number of elements... so it's not like you always pay with over 100% overhead until you reach 5000 elements.

**Comment by morder** [comment]
@finnschiermer but my simple test with 1000 rows gets 6.4-7.4 times more space used for DB. Compacted size i think it's a expected size of db with one thread writer.

_DB files size: Test: 786 432 
DB files size: Test: 122 880 (after compact)
... restart...
DB files size: Test: 917 504 
DB files size: Test: 122 880 (after compact)_
Another question, why DB size is different after each start? Do you know why this happens? 
Can I expect in future, db is not growing much more times? (10-20x+)

**Comment by finnschiermer** [comment]
Let me re-iterate: I think we are digressing. The original problem was that users reported to @morder about realm files of 300MB. I don't think the example project/code we are now discussing explains the original problem.

**Comment by morder** [comment]
@finnschiermer yes it's true, and I make a simple test to reproduce a problem, and I don't understand whats happens here...

**Comment by finnschiermer** [comment]
@morder: As the database grows, the overhead will become smaller, as the numbers illustrate. For example, if I add 20.000 objects it takes 3,145MByte before compact() and 2,621MByte after compact().

**Comment by finnschiermer** [comment]
In order to figure out what the original problem is (the 300MByte) we'll likely need the realm file or, if possible, a project which creates it.

**Comment by kneth** [comment]
@morder I am not sure what you mean by "questions with simple transaction model" but I hope that https://realm.io/docs/java/latest/#threading can ask them.

**Comment by diegomontoya** [comment]
@finnschiermer 

> so repackaging involves both allocation and freeing but always such that we allocate first, then commit, then recycle memory in the following transaction.

Everything is as expected, except the that the memory recycling of previous transaction's memory in next transaction is not really taking place here with @morder's test case due to severe fragmentation. What worries me is the "chunk/block" size you referred to in repacking the row which caused this fragmentation. Is this repacking chunk size fixed or dynamic? If you can point me to the relevant core file which has this size calculated/set, I will stop asking questions. =)

Reason I ask is that I like to optimize VM size over speed knowing my dbs are always going to remain small, less than 10K in element size so any fragmentation above a certain rate only means more potential OOM. I understand that fragmentation varies from wildly but I can still assume it is non-trivial based on this discussion.

**Comment by diegomontoya** [comment]
@morder Based on @finnschiermer's response, I think you have a compounded issue there. The fragmentation due to realm code is causing your db to be much larger than you expectation in the beginning but the size explosion levels off, stop accelerating at a certain point, and at a point that is well below 300MB. However, you may also have too many multi-threaded reading/writing here and forgot to close the open realms promptly which caused the secondary explosion of files.

**Comment by finnschiermer** [comment]
@morder: I agree with @diegomontoya.

@diegomontoya When you say less than 10K in element size, I take it you mean less than 10.000 elements? It depends a lot on the fields and the data you put in them, I'm afraid. If you compact() your databases you can see if you have a problem similar to what we've discussed here.

**Comment by finnschiermer** [comment]
@diegomontoya It is not simple to pinpoint a specific place which affects this. We are both allocating and releasing many different chunks of varying sizes and it happens in many different places. Even when you only see a single kind of object being manipulated. The actual allocation of the space in the realm file takes place in src/realm/groupwriter.cpp, if you feel adventorous :-O

**Comment by morder** [comment]
Thanks for answers. 
I have solved problem, we have used a threads and we are not closing realm after thread is finished.
Well, after compact and closing all realms we have new problem with 
_java.lang.IllegalArgumentException: Illegal Argument: Invalid format of Realm file_

I post a problem to exist tread
https://github.com/realm/realm-java/issues/3264

**Comment by kneth** [comment]
@morder I'll close this issue. But it is a bit strange if `compactRealm()` ends up corrupting your Realm file. But let's take that in the other issue :-)

**Comment by erudonja1** [comment]
http://stackoverflow.com/questions/34217476/android-realm-large-file-size/42959238#42959238

**Comment by Zhuinden** [comment]
@erudonja1 technically `compactRealm()` doesn't actually compact (and returns false) if you have random open Realm instance somewhere that is unclosed, you can use `Realm.getGlobalInstanceCount()` to verify this

**Comment by erudonja1** [comment]
@Zhuinden yes, that's correct, but this solution doesn't depend on those situations. Try it :)

**Comment by kneth** [comment]
@erudonja1 If you remove or overwrite a Realm file with open instances, I am not sure what will happen.

**Comment by erudonja1** [comment]
@kneth I was thinking the same thing, but seems like there is no problems. Maybe that is next feature for Realm repository, to create function like "closeAllInstances()", where ever they are opened.

Try it, if it makes problem, tell me :)

**Comment by Zhuinden** [comment]
It's your responsibility to know where you open Realm instances. :|

**Comment by erudonja1** [comment]
Yes, but what if you use 2 or 3 instances of database(not only default one), like I do for example in my case. 
The thing is that you will open connection in some activity that is probably stayed in cache/stack. And then you will do some big-insert operation that will cause problems of versions/memory...
When you try to compact db, it will return false, because those connections stayed opened. 

What I'm trying to say is, what if you can't close all thise connections because "onDestroy" of some activity is not called yet?

You can't destroy all activites...

**Comment by Zhuinden** [comment]
> you will do some big-insert operation that will cause problems of versions/memory...

That only causes problems if you do it in N transactions instead of 1 transaction, and you have many non-looper background threads operating on Realm at the same time.

**Comment by erudonja1** [comment]
@Zhuinden Thanks, I will try that also, but I'm not very optimistic that it will resize large realm file.

Anyway, seems like you guys were right about opened connections in Realm. 
If there are any opened connection or results(RealmResults) in some active-activity and I do this thing with file, it will fall. :(

I working on another solution right now, will update you about this later.

**Comment by erudonja1** [comment]
@kneth @Zhuinden 
Possible solution for me is to move that process of compacting database to be called from activity(new-one or some home activity) that will guarantee me that all activities(opened connections) are closed, stack is cleared, so default method Realm.compact(RealmConfiguration) would work.

Still, my solution (that is mentioned from before) is good, and eventually will cause problems if it is called from wrong place(activity that has opened connection to database). But if that is not the case, it will be fine. 

Mine conclusion is that: if you are working with sensitive process (like remove database, big-insert, big-update, new database, compact database space, copy from database to database etc.)  , do it/start it from neutral place(activity) that will ensure you that all activities before(in stack) with opened connections or RealmResults **are closed**. 
Then methods like:
Realm.getLocalInstanceCount(db.getConfiguration());
Realm.getGlobalInstanceCount(db.getConfiguration());
should give results 0, and **Realm.compact(db.getConfiguration()) will work**. Database file will be much smaller.(as it should be)

Correct me if I'm wrong.
I hope this will help someone, to not spend to much time searching for similar solution.

**Comment by kneth** [comment]
@erudonja1 Your conclusion is right when it comes to compacting. Many apps will compact when they launch since that's the time where they can ensure other instances are open. You can [write a new copy](https://realm.io/docs/java/3.0.0/api/io/realm/Realm.html#writeCopyTo-java.io.File-) of your Realm (which will compact it too) but then you will have to use the new copy instead.

Big inserts or updates should be fine to do at any time. If you are concern that it takes too long for your UI thread, doing async transactions might be a possibility.

**Comment by erudonja1** [comment]
That's why I say, from neutral place. Most secure way...

**Comment by tricuongle** [comment]
Dear all, 

I use Realm for API services to store database, and I got problem with Realm database size. 
I don't know this problem before I finish my program. 

Because it is multi thread in API Service, run 24/24 so it increasing 50M for 1000 update (it should not increasing data) and none stop. 
Problem is many call to API service per second (Only Update Data), and Sevices can't stop to Lauch Database again for Compact data. 

Could you suggest solution for this issue. It take me a lot of time but not solution for me. 

Thank you.

**Comment by cmelchior** [comment]
Hi @tricuongle Please create a new issue with as much information as possible. Old issues is not the place to resolve this. Thanks.
